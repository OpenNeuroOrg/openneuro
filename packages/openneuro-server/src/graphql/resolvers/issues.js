import mongo from '../../libs/mongo.js'

/**
 * Issues resolver
 */
export const issues = dataset => {
  return mongo.collections.crn.issues
    .findOne({
      id: dataset.revision,
      datasetId: dataset.id,
    })
    .then(data => (data ? data.issues : null))
}

/**
 * Snapshot issues resolver
 */
export const snapshotIssues = snapshot => {
  const datasetId = snapshot.id.split(':')[0]
  return mongo.collections.crn.issues
    .findOne({
      id: snapshot.hexsha,
      datasetId,
    })
    .then(data => (data ? data.issues : null))
}
