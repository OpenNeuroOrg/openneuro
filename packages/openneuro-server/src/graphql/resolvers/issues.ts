import Issue from "../../models/issue"
import { datasetType } from "./datasetType"
import { revalidate } from "./validation.js"

/**
 * Issues resolver
 */
export const issues = async (dataset, _, { userInfo }) => {
  return Issue.findOne({
    id: dataset.revision,
    datasetId: dataset.id,
    // Match if we have no validatorMetadata or the correct 'legacy' / 'schema' value if we do
    $or: [
      { "validatorMetadata.validator": await datasetType(dataset) },
      { validatorMetadata: { $exists: false } },
    ],
  })
    .exec()
    .then((data) => {
      if (!data && userInfo) {
        // If no results were found, acquire a lock and run validation
        revalidate(
          null,
          { datasetId: dataset.id, ref: dataset.revision },
          { userInfo },
        )
      }
      return data ? data.issues : null
    })
}

/**
 * Snapshot issues resolver
 */
export const snapshotIssues = async (snapshot) => {
  const datasetId = snapshot.id.split(":")[0]
  return Issue.findOne({
    id: snapshot.hexsha,
    datasetId,
    // Match if we have no validatorMetadata or the correct 'legacy' / 'schema' value if we do
    $or: [
      { "validatorMetadata.validator": await datasetType(snapshot) },
      { validatorMetadata: { $exists: false } },
    ],
  })
    .exec()
    .then((data) => (data ? data.issues : null))
}

/**
 * Resolver implementation for legacy validator status summary
 */
export async function issuesStatus(datasetId, version) {
  const data = await Issue.findOne({
    id: version,
    datasetId,
  }).exec()
  if (data) {
    const warnings = data.issues.filter((issue) =>
      issue.severity === "warning"
    ).length
    const errors = data.issues.filter((issue) =>
      issue.severity === "error"
    ).length
    return {
      errors,
      warnings,
    }
  } else {
    return null
  }
}

/**
 * Draft specific issues status resolver
 */
export function issuesDraftStatus(dataset) {
  return issuesStatus(dataset.id, dataset.revision)
}

/**
 * Draft specific issues status resolver
 */
export function issuesSnapshotStatus(snapshot) {
  const datasetId = snapshot.id.split(":")[0]
  return issuesStatus(datasetId, snapshot.hexsha)
}
