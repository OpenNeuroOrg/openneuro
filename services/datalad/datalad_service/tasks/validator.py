import asyncio
import json
import logging
import os
import re

import requests

from datalad_service.config import GRAPHQL_ENDPOINT

logger = logging.getLogger('datalad_service.' + __name__)

LEGACY_VALIDATOR_VERSION = json.load(
    open('package.json'))['dependencies']['bids-validator']
DENO_VALIDATOR_VERSION = 'v1.14.8'

LEGACY_METADATA = {
    'validator': 'legacy',
    'version': LEGACY_VALIDATOR_VERSION
}
DENO_METADATA = {
    'validator': 'schema',
    'version': DENO_VALIDATOR_VERSION
}


def escape_ansi(text):
    ansi_escape = re.compile(r'(?:\x1B[@-_]|[\x80-\x9F])[0-?]*[ -/]*[@-~]')
    return ansi_escape.sub('', text)


async def setup_validator():
    """Install nodejs deps if they do not exist."""
    if not os.path.exists('./node_modules/.bin/bids-validator'):
        process = await asyncio.create_subprocess_exec('yarn')
        await process.wait()


async def run_and_decode(args, timeout, logger):
    """Run a subprocess and return the JSON output."""
    process = await asyncio.create_subprocess_exec(*args, stdout=asyncio.subprocess.PIPE, stderr=asyncio.subprocess.PIPE)
    try:
        await asyncio.wait_for(process.wait(), timeout=timeout)
    except asyncio.TimeoutError:
        logger.warning(f'Timed out while running `{" ".join(args)}`')
        process.kill()

    # Retrieve what we can from the process
    stdout, stderr = await process.communicate()

    # If we have a whole JSON document from stdout, then return it
    # Otherwise, log the error and return None
    try:
        return json.loads(escape_ansi(stdout.decode('utf-8')))
    except json.decoder.JSONDecodeError as err:
        logger.exception(err)
        logger.info(stdout)
        logger.error(stderr)


async def validate_dataset_call(dataset_path, ref, logger=logger):
    """
    Synchronous dataset validation.

    Runs the bids-validator process and installs node dependencies if needed.
    """
    await setup_validator()
    return await run_and_decode(
        ['./node_modules/.bin/bids-validator', '--json', '--ignoreSubjectConsistency', dataset_path],
        timeout=300,
        logger=logger,
    )


async def validate_dataset_deno_call(dataset_path, ref, logger=logger):
    """
    Synchronous dataset validation.

    Runs the deno bids-validator process.
    """
    return await run_and_decode(
        ['deno', 'run', '-A',
         f'https://deno.land/x/bids_validator@{DENO_VALIDATOR_VERSION}/bids-validator.ts',
         '--json', dataset_path],
        timeout=300,
        logger=logger
    )


def summary_mutation(dataset_id, ref, validator_output, validator_metadata):
    """
    Return the OpenNeuro mutation to update a dataset summary.
    """
    summary = validator_output['summary']
    summary['datasetId'] = dataset_id
    # Set the object id to the git sha256 ref
    summary['id'] = ref
    summary['validatorMetadata'] = validator_metadata
    return {
        'query': 'mutation ($summaryInput: SummaryInput!) { updateSummary(summary: $summaryInput) { id } }',
        'variables':
            {
                'summaryInput': summary,
            }
    }


def issues_mutation(dataset_id, ref, issues, validator_metadata):
    """
    Return the OpenNeuro mutation to update any validation issues.
    """
    for issue in issues:
        # Remove extra stats to keep collection size down
        if 'files' in issue:
            for f in issue['files']:
                if f.get('file'):
                    if f['file'].get('stats'):
                        del f['file']['stats']
    issues = {
        'datasetId': dataset_id,
        'id': ref,
        'issues': issues,
        'validatorMetadata': validator_metadata
    }
    return {
        'query': 'mutation ($issues: ValidationInput!) { updateValidation(validation: $issues) }',
        'variables':
            {
                'issues': issues,
            }
    }


async def validate_dataset(dataset_id, dataset_path, ref, cookies=None, user=''):
    validator_output = await validate_dataset_call(dataset_path, ref)
    if validator_output:
        if 'issues' in validator_output:
            all_issues = validator_output['issues']['warnings'] + validator_output['issues']['errors']
            r = requests.post(
                url=GRAPHQL_ENDPOINT, json=issues_mutation(dataset_id, ref, all_issues, LEGACY_METADATA), cookies=cookies)
            if r.status_code != 200 or 'errors' in r.json():
                raise Exception(r.text)
        if 'summary' in validator_output:
            r = requests.post(
                url=GRAPHQL_ENDPOINT, json=summary_mutation(dataset_id, ref, validator_output, LEGACY_METADATA), cookies=cookies)
            if r.status_code != 200 or 'errors' in r.json():
                raise Exception(r.text)

    # New schema validator second in case of issues
    validator_output_deno = await validate_dataset_deno_call(dataset_path, ref)
    if validator_output_deno:
        if 'issues' in validator_output_deno:
            r = requests.post(
                url=GRAPHQL_ENDPOINT, json=issues_mutation(dataset_id, ref, validator_output_deno['issues'], DENO_METADATA), cookies=cookies)
            if r.status_code != 200 or 'errors' in r.json():
                raise Exception(r.text)
        if 'summary' in validator_output_deno:
            r = requests.post(
                url=GRAPHQL_ENDPOINT, json=summary_mutation(dataset_id, ref, validator_output_deno, DENO_METADATA), cookies=cookies)
            if r.status_code != 200 or 'errors' in r.json():
                raise Exception(r.text)



