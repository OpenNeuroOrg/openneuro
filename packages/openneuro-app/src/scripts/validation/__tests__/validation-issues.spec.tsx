import React from "react"
import { render, screen } from "@testing-library/react"
import { DatasetIssues } from "@bids/validator/issues"
import { Issue, Issues } from "../validation-issues"

describe("Issue component", () => {
  it("renders one issue", () => {
    const datasetIssues = new DatasetIssues()
    datasetIssues.codeMessages.set(
      "JSON_KEY_RECOMMENDED",
      "A JSON file is missing a key listed as recommended.",
    )
    const issue = {
      code: "JSON_KEY_RECOMMENDED",
      location: "/dataset_description.json",
      rule: "rules.dataset_metadata.dataset_description",
      subCode: "DatasetType",
    }

    const { asFragment } = render(
      <Issue datasetIssues={datasetIssues} issue={issue} groupBy="code" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
})
describe("Issues component", () => {
  it("renders multiple issues", () => {
    const datasetIssues = new DatasetIssues()
    datasetIssues.codeMessages.set(
      "JSON_KEY_RECOMMENDED",
      "A JSON file is missing a key listed as recommended.",
    )
    const issue = {
      code: "JSON_KEY_RECOMMENDED",
      location: "/dataset_description.json",
      rule: "rules.dataset_metadata.dataset_description",
      subCode: "DatasetType",
    }
    datasetIssues.issues = [{ ...issue }, { ...issue }, { ...issue }, {
      ...issue,
    }]
    const { asFragment } = render(
      <Issues issues={datasetIssues} groupBy="code" />,
    )
    expect(asFragment()).toMatchSnapshot()
  })
  it("sorts issues by the groupBy argument", async () => {
    const datasetIssues = new DatasetIssues()
    datasetIssues.codeMessages.set(
      "JSON_KEY_RECOMMENDED",
      "A JSON file is missing a key listed as recommended.",
    )
    datasetIssues.codeMessages.set(
      "SIDECAR_KEY_RECOMMENDED",
      "A data file's JSON sidecar is missing a key listed as recommended.",
    )
    datasetIssues.issues = [{
      code: "JSON_KEY_RECOMMENDED",
      location: "/dataset_description.json",
      severity: "warning",
      rule: "rules.dataset_metadata.dataset_description",
      subCode: "DatasetType",
    }, {
      code: "JSON_KEY_RECOMMENDED",
      subCode: "GeneratedBy",
      location: "/dataset_description.json",
      severity: "warning",
      rule: "rules.dataset_metadata.dataset_description",
    }, {
      code: "SIDECAR_KEY_RECOMMENDED",
      subCode: "ManufacturersModelName",
      location: "/sub-01/anat/sub-01_T1w.nii.gz",
      severity: "warning",
      rule: "rules.sidecars.mri.MRIHardware",
    }]
    await render(
      <Issues issues={datasetIssues} groupBy="location" />,
    )
    const headings = await screen.getAllByRole("heading")
    expect(headings[0].textContent).toMatch("/dataset_description.json2 issues")
    expect(headings[1].textContent).toMatch(
      "/sub-01/anat/sub-01_T1w.nii.gz1 issues",
    )
  })
})
