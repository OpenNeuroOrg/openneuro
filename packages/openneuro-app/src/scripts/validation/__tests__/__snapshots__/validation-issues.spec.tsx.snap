// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Issue component > renders one issue 1`] = `
<DocumentFragment>
  <div
    class="em-body"
  >
    <div
      class="e-meta"
    >
      <label>
        /dataset_description.json
      </label>
    </div>
    <div
      class="e-meta"
    >
      <label>
        Rule:
      </label>
       rules.dataset_metadata.dataset_description
    </div>
    <div
      class="e-meta"
    >
      <label>
        Messages:
      </label>
      <p>
        A JSON file is missing a key listed as recommended.
      </p>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Issues component > renders multiple issues 1`] = `
<DocumentFragment>
  <div
    class=" on-accordion-wrapper"
  >
    <article
      class="plain accordion validation-error fade-in"
    >
      <div
        class="accordion-title "
        role="switch"
      >
        <span
          class="file-header"
        >
          <h4
            class="em-header"
          >
            <strong
              class="em-header"
            >
              JSON_KEY_RECOMMENDED
            </strong>
            <span
              class="file-issue-count"
            >
              4 issues
            </span>
          </h4>
        </span>
      </div>
      <div
        class="accordion-item  collapsed "
      >
        <div
          class="accordion-content"
        >
          <div
            class="em-body"
          >
            <div
              class="e-meta"
            >
              <label>
                /dataset_description.json
              </label>
            </div>
            <div
              class="e-meta"
            >
              <label>
                Rule:
              </label>
               rules.dataset_metadata.dataset_description
            </div>
            <div
              class="e-meta"
            >
              <label>
                Messages:
              </label>
              <p>
                A JSON file is missing a key listed as recommended.
              </p>
            </div>
          </div>
          <div
            class="em-body"
          >
            <div
              class="e-meta"
            >
              <label>
                /dataset_description.json
              </label>
            </div>
            <div
              class="e-meta"
            >
              <label>
                Rule:
              </label>
               rules.dataset_metadata.dataset_description
            </div>
            <div
              class="e-meta"
            >
              <label>
                Messages:
              </label>
              <p>
                A JSON file is missing a key listed as recommended.
              </p>
            </div>
          </div>
          <div
            class="em-body"
          >
            <div
              class="e-meta"
            >
              <label>
                /dataset_description.json
              </label>
            </div>
            <div
              class="e-meta"
            >
              <label>
                Rule:
              </label>
               rules.dataset_metadata.dataset_description
            </div>
            <div
              class="e-meta"
            >
              <label>
                Messages:
              </label>
              <p>
                A JSON file is missing a key listed as recommended.
              </p>
            </div>
          </div>
          <div
            class="em-body"
          >
            <div
              class="e-meta"
            >
              <label>
                /dataset_description.json
              </label>
            </div>
            <div
              class="e-meta"
            >
              <label>
                Rule:
              </label>
               rules.dataset_metadata.dataset_description
            </div>
            <div
              class="e-meta"
            >
              <label>
                Messages:
              </label>
              <p>
                A JSON file is missing a key listed as recommended.
              </p>
            </div>
          </div>
        </div>
      </div>
    </article>
  </div>
</DocumentFragment>
`;
